extension/opampextension/opamp_agent.go

```diff
@@ -0,0 +1,244 @@
+	if o.opampClient != nil {
+		o.logger.Debug("Stopping OpAMP client...")
+		err := o.opampClient.Stop(ctx)
+		return err
+	}
+	return nil
+}
+
+func newOpampAgent(cfg *Config, logger *zap.Logger, build component.BuildInfo, res pcommon.Resource) (*opampAgent, error) {
+	agentType := build.Command
+
+	sn, ok := res.Attributes().Get(semconv.AttributeServiceName)
+	if ok {
+		agentType = sn.AsString()
+	}
+
+	agentVersion := build.Version
+
+	sv, ok := res.Attributes().Get(semconv.AttributeServiceVersion)
+	if ok {
+		agentVersion = sv.AsString()
+	}
+
+	uid := ulid.Make()
+
+	sid, ok := res.Attributes().Get(semconv.AttributeServiceInstanceID)
+	if ok {
+		puid, err := ulid.Parse(sid.AsString())
+		if err != nil {
+			return nil, err
+		}
+		uid = puid
+	}
+
+	if cfg.InstanceUID != "" {
+		puid, err := ulid.Parse(cfg.InstanceUID)
+		if err != nil {
+			return nil, err
+		}
+		uid = puid
+	}
+
+	agent := &opampAgent{
+		cfg:             cfg,
+		logger:          logger,
+		agentType:       agentType,
+		agentVersion:    agentVersion,
+		instanceId:      uid,
+		effectiveConfig: localConfig, // TODO: Replace with https://github.com/open-telemetry/opentelemetry-collector/issues/6596
+	}
+
+	return agent, nil
+}
+
+func stringKeyValue(key, value string) *protobufs.KeyValue {
+	return &protobufs.KeyValue{
+		Key: key,
+		Value: &protobufs.AnyValue{
+			Value: &protobufs.AnyValue_StringValue{StringValue: value},
+		},
+	}
+}
+
+func (o *opampAgent) createAgentDescription() error {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return err
+	}
+
+	ident := []*protobufs.KeyValue{
+		stringKeyValue("service.instance.id", o.instanceId.String()),
+		stringKeyValue("service.name", o.agentType),
+		stringKeyValue("service.version", o.agentVersion),
+	}
+
+	nonIdent := []*protobufs.KeyValue{
+		stringKeyValue("os.arch", runtime.GOARCH),
+		stringKeyValue("os.family", runtime.GOOS),
+		stringKeyValue("host.name", hostname),
+	}
+
+	o.agentDescription = &protobufs.AgentDescription{
+		IdentifyingAttributes:    ident,
+		NonIdentifyingAttributes: nonIdent,
+	}
+
+	return nil
+}
+
+func (o *opampAgent) updateAgentIdentity(instanceId ulid.ULID) {
+	o.logger.Debug("OpAMP agent identity is being changed",
+		zap.String("old_id", o.instanceId.String()),
+		zap.String("new_id", instanceId.String()))
+	o.instanceId = instanceId
+}
+
+func (o *opampAgent) composeEffectiveConfig() *protobufs.EffectiveConfig {
+	return &protobufs.EffectiveConfig{
+		ConfigMap: &protobufs.AgentConfigMap{
+			ConfigMap: map[string]*protobufs.AgentConfigFile{
+				"": {Body: []byte(o.effectiveConfig)},
+			},
+		},
+	}
+}
+
+func (o *opampAgent) onMessage(ctx context.Context, msg *types.MessageData) {
+	if msg.AgentIdentification != nil {
+		instanceId, err := ulid.Parse(msg.AgentIdentification.NewInstanceUid)
+		if err != nil {
+			o.logger.Error("Failed to parse a new agent identity", zap.Error(err))
+		}
+		o.updateAgentIdentity(instanceId)
+	}
```

Comment:
nit: prefer to return early from simple case
```suggestion
	if msg.AgentIdentification == nil {
		return
	}
	instanceId, err := ulid.Parse(msg.AgentIdentification.NewInstanceUid)
	if err != nil {
		o.logger.Error("Failed to parse a new agent identity", zap.Error(err))
	}
	o.updateAgentIdentity(instanceId)
```